src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs

```diff
@@ -712,6 +729,90 @@ public bool ShouldTreatWarningAsError(string warningCode)
             // An empty set means all warnings are errors.
             return WarningsAsErrors.Count == 0 || WarningsAsErrors.Contains(warningCode);
         }
+
+        /// <summary>
+        /// Allocates shared CPU resources. Called by a task when it's about to do potentially multi-threaded/multi-process work.
+        /// </summary>
+        /// <param name="requestedCores">The number of cores the task wants to use.</param>
+        /// <returns>The number of cores the task is allowed to use given the current state of the build. This number is always between
+        /// 1 and <paramref name="requestedCores"/>. If the task has allocated its one implicit core, this call may block, waiting for
+        /// at least one core to become available.</returns>
+        public int RequestCores(int requestedCores)
+        {
+            ErrorUtilities.VerifyThrowArgumentOutOfRange(requestedCores > 0, nameof(requestedCores));
+
+            lock (_callbackMonitor)
+            {
+                IRequestBuilderCallback builderCallback = _requestEntry.Builder as IRequestBuilderCallback;
+
+                int coresAcquired = 0;
+                bool allocatingImplicitCore = false;
+                if (_isImplicitCoreUsed)
+                {
+                    coresAcquired = builderCallback.RequestCores(_callbackMonitor, requestedCores, waitForCores: true);
+                }
+                else
+                {
+                    _isImplicitCoreUsed = true;
+                    allocatingImplicitCore = true;
+                    if (requestedCores > 1)
+                    {
+                        coresAcquired = builderCallback.RequestCores(_callbackMonitor, requestedCores - 1, waitForCores: false);
```

Comment:
I'm assuming this always returns at least 1. Does that mean that if you're allocating the implicit core (for the task) that you're always going to get at least 2? And it might still block?

Reply:
Passing `waitForCores: false` makes 0 a possible return value from the builder callback. So this will return at least 1 (for the implicit core). Only if further calls are made do we take the other branch which blocks.